### PR TITLE
Update product's archive status

### PR DESCRIPTION
**Pull request by tdddblog**
## 🗒️ Summary
Update product's archive status - individual products, collections and primary references, bundles.

## ⚙️ Test Data and/or Report
One of the following should be included here:
* Reference to regression test included in code (preferred wherever reasonable)
* Attach test data here + outputs of tests

## ♻️ Related Issues
https://github.com/NASA-PDS/registry-harvest-cli/issues/2
https://github.com/nasa-pds/registry-harvest-service/issues/12

### CODE REVIEW

File: src/main/java/gov/nasa/pds/registry/common/es/dao/DaoUtils.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+     * @param is input stream
+     * @return Last line
+     */
+    public static String getLastLine(InputStream is)
+    {
+        String lastLine = null;
+
+        try
+        {
+            BufferedReader rd = new BufferedReader(new InputStreamReader(is));
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/DaoUtils.java:35]: ignore

File: src/main/java/gov/nasa/pds/registry/common/meta/FileMetadataExtractor.java [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+        // Zipped content holder
+        ByteArrayOutputStream bas = new ByteArrayOutputStream();
+        // Zipper
+        DeflaterOutputStream deflater = new DeflaterOutputStream(bas);
+        // Writer to output stream adapter
+        OutputStreamWriter dest = new OutputStreamWriter(deflater);
+        
+        try
+        {
+            source = new FileReader(file);
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/meta/FileMetadataExtractor.java:255]: ignore

File: src/main/java/gov/nasa/pds/registry/common/meta/FileMetadataExtractor.java [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    public static String getJsonBlob(File file) throws Exception
+    {
+        Reader source = null;
+        
+        // Zipped content holder
+        ByteArrayOutputStream bas = new ByteArrayOutputStream();
+        // Zipper
+        DeflaterOutputStream deflater = new DeflaterOutputStream(bas);
+        // Writer to output stream adapter
+        OutputStreamWriter dest = new OutputStreamWriter(deflater);
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/meta/FileMetadataExtractor.java:251]: ignore

File: src/main/java/gov/nasa/pds/registry/common/es/dao/ProductDao.java [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+                throw ex;
+            }
+        }
+
+        InputStream is = null;
+
+        try
+        {
+            is = resp.getEntity().getContent();
+            JsonReader rd = new JsonReader(new InputStreamReader(is));
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/ProductDao.java:172]: ignore

File: src/main/java/gov/nasa/pds/registry/common/meta/CollectionMetadataExtractor.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ * @author karpenko
+ */
+public class CollectionMetadataExtractor
+{
+    private XPathExpression xFileName;
+    
+
+    /**
+     * Constructor
+     * @throws Exception
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/meta/CollectionMetadataExtractor.java]: fixed

File: src/main/java/gov/nasa/pds/registry/common/meta/FieldNameCache.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    public static FieldNameCache getInstance()
+    {
+        return singleton;
+    }
+    
+    
+    /**
+     * Set cached values
+     * @param fieldNames
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/meta/FieldNameCache.java]: fixed

File: src/main/java/gov/nasa/pds/registry/common/meta/Metadata.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    public Metadata()
+    {
+        intRefs = new FieldMapSet();
+        fields = new FieldMapList();
+    }
+    
+
+    /**
+     * Set node name
+     * @param name
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/meta/Metadata.java]: fixed

File: src/main/java/gov/nasa/pds/registry/common/meta/Metadata.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+     */
+    public void setNodeName(String name)
+    {
+        fields.setValue(FLD_NODE_NAME, name);
+    }
+    
+    
+    /**
+     * Set harvest timestamp
+     * @param val
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/meta/Metadata.java]: fixed

File: src/main/java/gov/nasa/pds/registry/common/meta/Metadata.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    public void setHarvestTimestamp(Instant val)
+    {
+        String strVal = DateTimeFormatter.ISO_INSTANT.format(val);
+        fields.setValue(FLD_HARVEST_DATE_TIME, strVal);
+    }
+
+
+    /**
+     * Set archive status
+     * @param name
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/meta/Metadata.java]: fixed

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    {
+        Collection<String> values = getValues(fieldName);
+        return (values == null || values.isEmpty()) ? null : values.iterator().next();
+    }
+
+
+    /**
+     * Get all values of a field.
+     */
+    public Collection<String> getValues(String fieldName)
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java:124]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    {
+        return fields.get(fieldName);
+    }
+
+    
+    /**
+     * Get names of all fields in this map.
+     */
+    public Set<String> getNames()
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java:133]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        
+        List<String> set = getOrCreateValues(fieldName);        
+        Collections.addAll(set, values);
+    }
+
+
+    /**
+     * Get first value of a field.
+     */
+    public String getFirstValue(String fieldName)
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java:114]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    public int size()
+    {
+        return fields.size();
+    }
+    
+    
+    /**
+     * Check if map is empty.
+     */
+    public boolean isEmpty()
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java:45]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+    public FieldMapList()
+    {
+        fields = new TreeMap<>();
+    }
+    
+    
+    /**
+     * Map size / number of fields.
+     */
+    public int size()
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java:36]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+        if(value == null) return;
+        
+        fields.put(fieldName, Arrays.asList(value));
+    }
+    
+    
+    /**
+     * Add field's value
+     */
+    public void addValue(String fieldName, String value)
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java:90]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+        
+        List<String> values = getOrCreateValues(fieldName);        
+        values.add(value);
+    }
+    
+
+    /**
+     * Add multiple values
+     */
+    public void addValues(String fieldName, String[] values)
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java:102]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+    public boolean isEmpty()
+    {
+        return fields.size() == 0;
+    }
+    
+    
+    /**
+     * Remove all the fields from the map
+     */
+    public void clear()
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapList.java:54]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapSet.java [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        Collection<String> values = getValues(fieldName);
+        return (values == null || values.isEmpty()) ? null : values.iterator().next();
+    }
+
+
+    /**
+     * Get all values of a field.
+     */
+    public Collection<String> getValues(String fieldName)
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapSet.java:110]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapSet.java [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+        
+        Set<String> set = getOrCreateValues(fieldName);        
+        Collections.addAll(set, values);
+    }
+
+
+    /**
+     * Get first value of a field.
+     */
+    public String getFirstValue(String fieldName)
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapSet.java:100]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapSet.java [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+    public int size()
+    {
+        return fields.size();
+    }
+    
+    
+    /**
+     * Check if map is empty.
+     */
+    public boolean isEmpty()
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapSet.java:43]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapSet.java [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+    public FieldMapSet()
+    {
+        fields = new TreeMap<>();
+    }
+    
+    
+    /**
+     * Map size / number of fields.
+     */
+    public int size()
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapSet.java:34]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapSet.java [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        return values;
+    }
+    
+    
+    /**
+     * Add field's value.
+     */
+    public void addValue(String fieldName, String value)
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapSet.java:76]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapSet.java [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+        
+        Set<String> values = getOrCreateValues(fieldName);        
+        values.add(value);
+    }
+    
+
+    /**
+     * Add multiple values.
+     */
+    public void addValues(String fieldName, String[] values)
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapSet.java:88]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/FieldMapSet.java [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+    public boolean isEmpty()
+    {
+        return fields.size() == 0;
+    }
+    
+    
+    /**
+     * Remove all the fields from the map
+     */
+    public void clear()
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/FieldMapSet.java:52]: ignore

File: src/main/java/gov/nasa/pds/registry/common/meta/BundleMetadataExtractor.java [hunk 25]
```diff
@@ -0,0 +1,153 @@
...
+                bme.lid = node.getTextContent().trim();
+            }
+            else if(nodeName.equals("lidvid_reference"))
+            {
+                bme.lidvid = node.getTextContent().trim();
+            }
+            else if(nodeName.equals("reference_type"))
+            {
+                bme.type = node.getTextContent().trim();
+                String[] tokens = bme.type.split("_");
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/meta/BundleMetadataExtractor.java:138]: ignore

File: src/main/java/gov/nasa/pds/registry/common/meta/InternalReferenceExtractor.java [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+        {
+            String key = "ref_lidvid_" + type;
+            fmap.addValue(key, ref.lidvid);
+        }
+    }
+    
+    
+    private String getShortRefType(String refType)
+    {
+        String[] tokens = refType.split("_");
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/meta/InternalReferenceExtractor.java:88]: ignore

File: src/main/java/gov/nasa/pds/registry/common/meta/BasicMetadataExtractor.java [hunk 27]
```diff
@@ -0,0 +1,167 @@
...
+            {
+                dir = node.getTextContent().trim();
+            }
+        }
+
+        if(fileName == null) return null;
+
+        if(dir == null) return fileName;
+        
+        return (dir.endsWith("/")) ? dir + fileName : dir + "/" + fileName;
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/meta/BasicMetadataExtractor.java]: fixed

File: pom.xml [hunk 28]
```diff
@@ -104,10 +110,29 @@ POSSIBILITY OF SUCH DAMAGE.
             <artifactId>commons-lang3</artifactId>
             <version>3.12.0</version>
         </dependency>
+        <!-- File type detection -->
+        <dependency>
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply pom.xml:114]: ignore

File: src/main/java/gov/nasa/pds/registry/common/meta/BasicMetadataExtractor.java [hunk 29]
```diff
@@ -0,0 +1,167 @@
...
+    private XPathExpression xLid;
+    private XPathExpression xVid;
+    private XPathExpression xTitle;
+
+    private XPathExpression xFileName;
+    private XPathExpression xDocFile;
+    
+    /**
+     * Constructor
+     * @param config configuration
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/meta/BasicMetadataExtractor.java]: fixed

**Review by tloubrieu-jpl — APPROVED** [review]
That works, thanks

**Pull request merged**